icicle/src/merkle-tree/mmcs.cu

```diff
@@ -0,0 +1,458 @@
+
+      if (params.number_of_leaves_to_inject) {
+        std::cout << "Injecting " << params.number_of_rows << std::endl;
+        CHK_IF_RETURN(params.compression->compress_and_inject(
+          d_leaves_info, params.number_of_leaves_to_inject, params.number_of_rows, states, digests,
+          params.digest_elements, *params.ctx));
+        std::cout << "Injected" << std::endl;
+      } else {
+        std::cout << "Compressing" << std::endl;
+        CHK_IF_RETURN(params.compression->compress_many(
+          states, digests, params.number_of_rows_padded, params.digest_elements, *params.ctx));
+        std::cout << "Compressed" << std::endl;
+      }
+
+      CHK_IF_RETURN(maybe_copy_digests(digests, big_tree_digests, params));
+      swap<D>(&digests, &states);
+      params.segment_size /= params.arity;
+      params.subtree_height--;
+      params.number_of_rows_padded /= params.arity;
+    }
+
+    return CHK_LAST();
+  }
+
+  template <typename L, typename D>
+  cudaError_t build_mmcs_subtree(
+    const std::vector<Matrix<L>>& leaves,
+    L* d_leaves,
+    D* states,
+    L* aux_leaves_mem,
+    L* big_tree_digests,
+    SubtreeParams<L, D>& params)
+  {
+    // Leaves info
+    Matrix<L>* d_leaves_info;
+    CHK_IF_RETURN(cudaMallocAsync(&d_leaves_info, params.number_of_inputs * sizeof(Matrix<L>), params.ctx->stream));
+
+    CHK_IF_RETURN(slice_and_copy_leaves(leaves, d_leaves, d_leaves_info, params));
+
+    // Reuse leaves memory
+    D* digests = (D*)d_leaves;
+
+    std::cout << "Hashing bottom layer leaves " << params.number_of_rows << std::endl;
+    CHK_IF_RETURN(hash_leaves(
+      d_leaves_info, params.number_of_leaves_to_inject, params.number_of_rows, states, digests, params.digest_elements,
+      *params.hasher, *params.ctx));
+    std::cout << "Hashed bottom layer leaves" << std::endl;
+
+    CHK_IF_RETURN(maybe_copy_digests(digests, big_tree_digests, params));
+
+    params.number_of_rows_padded /= params.arity;
+    params.segment_size /= params.arity;
+    params.subtree_height--;
+
+    swap<D>(&digests, &states);
+    return fold_layers<L, D>(leaves, states, digests, aux_leaves_mem, d_leaves_info, big_tree_digests, params);
+  }
+
+  template <typename L, typename D>
+  cudaError_t mmcs_commit(
+    const Matrix<L>* inputs,
+    const unsigned int number_of_inputs,
+    D* digests,
+    const SpongeHasher<L, D>& hasher,
+    const SpongeHasher<L, D>& compression,
+    const TreeBuilderConfig& tree_config)
+  {
+    CHK_INIT_IF_RETURN();
+    cudaStream_t& stream = tree_config.ctx.stream;
+
+    if (number_of_inputs == 0) THROW_ICICLE_ERR(IcicleError_t::InvalidArgument, "No matrices provided");
+
+    if (compression.preimage_max_length < tree_config.arity * tree_config.digest_elements)
+      THROW_ICICLE_ERR(
+        IcicleError_t::InvalidArgument,
+        "Hash max preimage length does not match merkle tree arity multiplied by digest elements");
+
+    std::vector<Matrix<L>> sorted_inputs(number_of_inputs);
```

Comment:
might be faster to use a list rather than a vector for the sorting if you have a lot of inputs